pyterrier/model.py

```diff
@@ -85,80 +85,134 @@ def query_columns(df : pd.DataFrame, qid=True) -> Sequence[str]:
-                supplied dataframe. Defaults to False. 
+    Changes a dataframe such that the selected column becomes "<column>_0", and any
+    "<column>_0" columns becomes "<column>_1" etc.
+
+    Arguments:
+        df: Dataframe with a "<column>" column
+        keep_original: if True, the <column> column is also left unchanged. Useful for client code.
+            Defaults to False.
+        inplace: if False, a copy of the dataframe is returned. If True, changes are made to the
+            supplied dataframe. Defaults to False.
     """
     cols = set(df.columns)
-    if "query" not in cols:
-        raise KeyError(f"Expected a query column, but found {list(cols)}")
+    if base_column not in cols:
+        raise KeyError(f"Expected a {base_column} column, but found {list(cols)}")
     if not inplace:
         df = df.copy()
-    prev_col = 'query'
+    prev_col = base_column
     rename_cols = {}
     for query_idx in itertools.count():
-        next_col = f'query_{query_idx}'
+        next_col = f"{base_column}_{query_idx}"
         if prev_col in cols:
-            rename_cols[prev_col] = next_col # map e.g., query_0 to be renamed to query_1
+            rename_cols[prev_col] = next_col
             prev_col = next_col
         else:
-             break
-    df = df.rename(columns=rename_cols)
+            break
+    # apply renaming in-place or on the copy
+    df.rename(columns=rename_cols, inplace=True)
     if keep_original:
-        df['query'] = df["query_0"]
+        df[base_column] = df[f"{base_column}_0"]
     return df
 
-
-def push_queries_dict(inp: IterDictRecord, *, keep_original: bool = False, inplace: bool = False) -> IterDictRecord:
+@overload
+def push_columns_dict(
+    inp: IterDict,
+    keep_original: bool = False,
+    base_column: str = "query",
+) -> IterDict: ...
+
+@overload
+def push_columns_dict(
+    inp: IterDictRecord,
+    keep_original: bool = False,
+    base_column: str = "query",
+) -> IterDictRecord: ...
+
+def push_columns_dict(
+    inp: Union[IterDict, IterDictRecord],
+    keep_original: bool = False,
+    base_column: str = "query",
+) -> Union[IterDict, IterDictRecord]:
     """
-    Works like ``push_queries`` but over a dict instead of a dataframe.
+    Works like ``push_columns`` but over a dict/iterdict instead of a dataframe.
     """
-    if "query" not in inp:
-        raise KeyError(f"Expected a query column, but found {list(inp.keys())}")
-    if not inplace:
-        inp = inp.copy()
-    prev_col = 'query'
-    for query_idx in itertools.count():
-        next_col = f'query_{query_idx}'
-        if prev_col in inp:
-            inp[next_col] = inp.pop(prev_col) # assign e.g., query_1 to query_0
-            prev_col = next_col
-        else:
-             break
-    if keep_original:
-        inp['query'] = inp['query_0']
-    return inp
+    def per_element(i: IterDictRecord) -> IterDictRecord:
```

Comment:
We may want to try to make this a bit more efficient in the `IterDict` case, since it's doing a lot of work for every record. But it's not a regression, so not necessary to address in this PR.